File: examples/backends/typescript/test/api.test.ts
/**
 * End-to-end checks that need no Meteroid tenant.
 *
 * Everything reachable without an API key is exercised here against the shapes
 * `examples/openapi.yaml` promises: the health probe, the session-token gate, the error
 * envelope, the 404 fallback, and — the interesting one — webhook signature
 * verification, which works offline because the SDK ships a *signer* as well as a
 * verifier. The contract suite uses the same trick to test the receiver for real.
 */

import assert from "node:assert/strict";
import { test } from "node:test";

import { MAX_BODY_BYTES } from "../src/http.js";
import { mint } from "../src/session.js";
import { AppState } from "../src/state.js";
import { assertErrorEnvelope, CONFIG, fakeMeteroid, SESSION_SECRET, send, signed } from "./support.js";

const bearer = (alias = "scribe-demo-1") => ({
  authorization: `Bearer ${mint(SESSION_SECRET, alias)}`,
  "content-type": "application/json",
});

test("health reports the backend and its configuration", async () => {
  const { status, body } = await send("GET", "/api/health");

  assert.equal(status, 200);
  assert.deepEqual(Object.keys(body).sort(), ["backend", "meteroid_configured", "status", "version"]);
  assert.equal(body.status, "ok");
  assert.equal(body.backend, "typescript");
  assert.equal(body.meteroid_configured, false);
  assert.equal(typeof body.version, "string");
});

test("a missing session token is 401 in the standard envelope", async () => {
  const { status, body } = await send("GET", "/api/me");

  assert.equal(status, 401);
  assertErrorEnvelope(body, "UNAUTHORIZED");
});

test("a token from another deployment is rejected", async () => {
  const forged = mint("some-other-secret", "scribe-demo-1");
  const { status, body } = await send("GET", "/api/me", {
    headers: { authorization: `Bearer ${forged}` },
  });

  assert.equal(status, 401);
  assertErrorEnvelope(body, "UNAUTHORIZED");
});

test("an Authorization header that is not a bearer token is rejected", async () => {
  for (const authorization of ["Basic abc", `bearer ${mint(SESSION_SECRET, "a")}`, "Bearer ", "Bearer garbage"]) {
    const { status, body } = await send("GET", "/api/me", { headers: { authorization } });

    assert.equal(status, 401, authorization);
    assertErrorEnvelope(body, "UNAUTHORIZED");
  }
});

test("the session token is checked before the body is looked at", async () => {
  const { status, body } = await send("POST", "/api/transcriptions", { body: "{not json" });

  assert.equal(status, 401);
  assertErrorEnvelope(body, "UNAUTHORIZED");
});

test("an unknown route is 404 in the standard envelope", async () => {
  for (const path of ["/api/nope", "/", "/api/health/", "/api/__proto__"]) {
    const { status, body } = await send("GET", path);

    assert.equal(status, 404, path);
    assertErrorEnvelope(body, "NOT_FOUND");
  }
});

test("a known route with the wrong method is a bare 405 naming what is allowed", async () => {
  const wrong = await send("DELETE", "/api/health");
  assert.equal(wrong.status, 405);
  assert.equal(wrong.headers["allow"], "GET,HEAD");
  assert.equal(wrong.body, null);

  const both = await send("PUT", "/api/transcriptions");
  assert.equal(both.headers["allow"], "GET,HEAD,POST");
});

test("every answer carries permissive CORS headers, and OPTIONS is a preflight", async () => {
  const simple = await send("GET", "/api/nope");
  assert.equal(simple.headers["access-control-allow-origin"], "*");

  const preflight = await send("OPTIONS", "/api/checkout", {
    headers: { origin: "http://localhost:5173", "access-control-request-method": "POST" },
  });
  assert.equal(preflight.status, 200);
  assert.equal(preflight.headers["access-control-allow-origin"], "*");
  assert.equal(preflight.headers["access-control-allow-methods"], "*");
  assert.equal(preflight.headers["access-control-allow-headers"], "*");
});

test("a malformed body is 400 in the standard envelope", async () => {
  const bodies = [
    // `duration_seconds` is required, and unknown fields are rejected.
    '{"title":"x","surprise":true}',
    '{"title":"x"}',
    "{not json",
    "[]",
    "null",
    "",
    "   ",
    // Wrong scalar types are type errors, never coerced.
    '{"title":123,"duration_seconds":60}',
    '{"title":"x","duration_seconds":"60"}',
    '{"title":"x","duration_seconds":60.5}',
    '{"title":"x","duration_seconds":true}',
    '{"title":null,"duration_seconds":60}',
    // Well-typed, but outside the contract's bounds.
    '{"title":"x","duration_seconds":0}',
    '{"title":"x","duration_seconds":7201}',
    '{"title":"   ","duration_seconds":60}',
    `{"title":"${"x".repeat(201)}","duration_seconds":60}`,
  ];
  for (const raw of bodies) {
    const meteroid = fakeMeteroid();
    const { status, body } = await send("POST", "/api/transcriptions", {
      headers: bearer(),
      body: raw,
      meteroid: meteroid.client,
    });

    assert.equal(status, 400, raw);
    assertErrorEnvelope(body, "BAD_REQUEST");
    assert.deepEqual(meteroid.calls, [], `Meteroid was called for ${raw}`);
  }
});

test("a body that is not valid UTF-8 is 400, not a replacement character", async () => {
  const { status, body } = await send("POST", "/api/session", {
    body: Buffer.from([0x7b, 0x22, 0xff, 0x22, 0x3a, 0x31, 0x7d]),
  });

  assert.equal(status, 400);
  assertErrorEnvelope(body, "BAD_REQUEST");
});

test("an optional body rejects what it does not declare", async () => {
  for (const raw of ['{"nope":1}', '{"workspace_name":123}', '{"workspace_name":"  "}', '{"email":false}', "[]"]) {
    const meteroid = fakeMeteroid();
    const { status, body } = await send("POST", "/api/session", { body: raw, meteroid: meteroid.client });

    assert.equal(status, 400, raw);
    assertErrorEnvelope(body, "BAD_REQUEST");
    assert.deepEqual(meteroid.calls, []);
  }
});

test("a plan code outside the contract enum is rejected before the catalog is consulted", async () => {
  const meteroid = fakeMeteroid();
  const { status, body } = await send("POST", "/api/checkout", {
    headers: bearer(),
    body: '{"plan_code":"gold"}',
    meteroid: meteroid.client,
  });

  assert.equal(status, 400);
  assertErrorEnvelope(body, "BAD_REQUEST");
  assert.deepEqual(meteroid.calls, []);
});

test("an out-of-range portal lifetime is rejected before Meteroid sees it", async () => {
  for (const raw of ['{"expires_in_seconds":5}', '{"expires_in_seconds":2592001}', '{"expires_in_seconds":"60"}']) {
    const meteroid = fakeMeteroid();
    const { status, body } = await send("POST", "/api/portal-session", {
      headers: bearer(),
      body: raw,
      meteroid: meteroid.client,
    });

    assert.equal(status, 400, raw);
    assertErrorEnvelope(body, "BAD_REQUEST");
    assert.deepEqual(meteroid.calls, []);
  }
});

test("an invalid invoice limit is rejected before Meteroid sees it", async () => {
  for (const query of ["limit=0", "limit=101", "limit=abc", "limit=", "limit=1.5", "foo=1", "limit=1&limit=2"]) {
    const meteroid = fakeMeteroid();
    const { status, body } = await send("GET", `/api/invoices?${query}`, {
      headers: bearer(),
      meteroid: meteroid.client,
    });

    assert.equal(status, 400, query);
    assertErrorEnvelope(body, "BAD_REQUEST");
    assert.deepEqual(meteroid.calls, []);
  }
});

test("an oversized body is refused in the standard envelope", async () => {
  // `null` is how `http.ts` reports a body it stopped buffering at the limit.
  const { status, body } = await send("POST", "/api/session", { body: null });

  assert.equal(status, 413);
  assertErrorEnvelope(body, "BAD_REQUEST");
  assert.ok(MAX_BODY_BYTES > 0);
});

test("an unreachable Meteroid is 502, not a crash", async () => {
  // No fake here: the real client, pointed at a port nothing listens on.
  const state = new AppState({ ...CONFIG, meteroidBaseUrl: "http://127.0.0.1:9" });
  const { status, body } = await send("GET", "/api/plans", { state });

  assert.equal(status, 502);
  assertErrorEnvelope(body, "UPSTREAM_ERROR");
});

// ---------------------------------------------------------------- webhooks

test("a correctly signed event is accepted", async () => {
  const payload = JSON.stringify({ id: "evt_123", type: "invoice.paid", timestamp: "2026-09-01T12:00:00Z" });
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: signed(payload, "msg_1"),
    body: payload,
  });

  assert.equal(status, 202);
  assert.deepEqual(body, { received: true, event_id: "evt_123", type: "invoice.paid", handled: true });
});

test("the raw bytes are verified, not a re-serialization of them", async () => {
  // Spacing and key order no JSON serializer would reproduce.
  const payload = '{ "type" : "invoice.paid",\n\t"id":"evt_raw"   }';
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: signed(payload, "msg_raw"),
    body: payload,
  });

  assert.equal(status, 202);
  assert.equal(body.event_id, "evt_raw");
});

test("an unknown event type is acknowledged, not rejected", async () => {
  // A receiver that 400s on an unrecognized type breaks the first time Meteroid ships a
  // new one — so this must be a 202 with `handled: false`. The body also carries no
  // `timestamp`, which must not matter either.
  const payload = JSON.stringify({ id: "evt_9", type: "something.invented.later" });
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: signed(payload, "msg_9"),
    body: payload,
  });

  assert.equal(status, 202);
  assert.equal(body.handled, false);
  assert.equal(body.event_id, "evt_9");
});

test("a signed body of an unexpected shape is acknowledged too", async () => {
  for (const payload of ["[1,2]", '{"id":7,"type":["x"]}', '"just a string"']) {
    const { status, body } = await send("POST", "/api/webhooks/meteroid", {
      headers: signed(payload, "msg_shape"),
      body: payload,
    });

    assert.equal(status, 202, payload);
    assert.deepEqual(body, { received: true, event_id: "msg_shape", type: null, handled: false });
  }
});

test("an event with no id falls back to the webhook-id header", async () => {
  const payload = JSON.stringify({ type: "subscription.created" });
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: signed(payload, "msg_42"),
    body: payload,
  });

  assert.equal(status, 202);
  assert.equal(body.event_id, "msg_42");
  assert.equal(body.handled, true);
});

test("svix headers are accepted as aliases", async () => {
  const payload = JSON.stringify({ type: "invoice.finalized" });
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: signed(payload, "msg_s", "svix"),
    body: payload,
  });

  assert.equal(status, 202);
  // The id fallback has to follow the alias as well.
  assert.equal(body.event_id, "msg_s");
});

test("a body that does not match its signature is rejected", async () => {
  const payload = JSON.stringify({ id: "evt_1", type: "invoice.paid" });
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: signed(payload, "msg_1"),
    // One field different from what was signed.
    body: JSON.stringify({ id: "evt_1", type: "invoice.void" }),
  });

  assert.equal(status, 400);
  assertErrorEnvelope(body, "WEBHOOK_SIGNATURE_INVALID");
});

test("a stale timestamp is rejected", async () => {
  const payload = JSON.stringify({ type: "invoice.paid" });
  const headers = signed(payload, "msg_old");
  headers["webhook-timestamp"] = String(Math.floor(Date.now() / 1000) - 3600);
  const { status, body } = await send("POST", "/api/webhooks/meteroid", { headers, body: payload });

  assert.equal(status, 400);
  assertErrorEnvelope(body, "WEBHOOK_SIGNATURE_INVALID");
});

test("an unsigned event is rejected", async () => {
  const { status, body } = await send("POST", "/api/webhooks/meteroid", {
    headers: { "content-type": "application/json" },
    body: '{"type":"invoice.paid"}',
  });

  assert.equal(status, 400);
  assertErrorEnvelope(body, "WEBHOOK_SIGNATURE_INVALID");
});

test("a correctly signed body that is not JSON is rejected", async () => {
  for (const payload of ["not json", ""]) {
    const { status, body } = await send("POST", "/api/webhooks/meteroid", {
      headers: signed(payload, "msg_nj"),
      body: payload,
    });

    assert.equal(status, 400, JSON.stringify(payload));
    assertErrorEnvelope(body, "WEBHOOK_SIGNATURE_INVALID");
  }
});

test("a missing webhook secret is the operator's problem, not a bad signature", async () => {
  const state = new AppState({ ...CONFIG, meteroidWebhookSecret: "" }, fakeMeteroid().client);
  const { status, body } = await send("POST", "/api/webhooks/meteroid", { state, body: "{}" });

  assert.equal(status, 500);
  assertErrorEnvelope(body, "INTERNAL");
});

File: examples/backends/go/api_test.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"testing"
	"time"
)

// End-to-end checks that need no Meteroid tenant.
//
// Everything reachable without an API key is exercised here against the shapes
// examples/openapi.yaml promises: the health probe, the session-token gate, the error
// envelope, the 404 fallback, and — the interesting one — webhook signature
// verification, which works offline because the SDK ships a *signer* as well as a
// verifier. The contract suite uses the same trick to test the receiver for real.

func TestHealthReportsTheBackendAndItsConfiguration(t *testing.T) {
	response := send(t, "GET", "/api/health", sendOptions{})

	if response.status != 200 {
		t.Fatalf("status = %d", response.status)
	}
	assertJSON(t, response.body, `{"status":"ok","backend":"go","meteroid_configured":false,"version":"1.0.0"}`)
	if got := response.header.Get("Content-Type"); got != "application/json" {
		t.Errorf("Content-Type = %q", got)
	}
}

func TestMissingSessionTokenIs401(t *testing.T) {
	response := send(t, "GET", "/api/me", sendOptions{})
	assertErrorEnvelope(t, response, 401, "UNAUTHORIZED")
	// Not "<session_token>": the message is meant to be read.
	if !strings.Contains(response.raw, "Bearer <session_token>") {
		t.Errorf("message was HTML-escaped: %s", response.raw)
	}
}

func TestTokenFromAnotherDeploymentIsRejected(t *testing.T) {
	forged := mintToken("some-other-secret", "scribe-demo-1")
	response := send(t, "GET", "/api/me", sendOptions{headers: map[string]string{"Authorization": "Bearer " + forged}})
	assertErrorEnvelope(t, response, 401, "UNAUTHORIZED")
}

func TestAuthorizationThatIsNotABearerTokenIsRejected(t *testing.T) {
	for _, authorization := range []string{"Basic abc", "bearer " + mintToken(testSessionSecret, "a"), "Bearer ", "Bearer garbage"} {
		response := send(t, "GET", "/api/me", sendOptions{headers: map[string]string{"Authorization": authorization}})
		assertErrorEnvelope(t, response, 401, "UNAUTHORIZED")
	}
}

func TestSessionTokenIsCheckedBeforeTheBody(t *testing.T) {
	response := send(t, "POST", "/api/transcriptions", sendOptions{body: "{not json"})
	assertErrorEnvelope(t, response, 401, "UNAUTHORIZED")

	oversized := send(t, "POST", "/api/transcriptions", sendOptions{body: strings.Repeat("x", maxBodyBytes+1)})
	assertErrorEnvelope(t, oversized, 401, "UNAUTHORIZED")
}

func TestUnknownRouteIs404(t *testing.T) {
	// A mux would redirect the second and third of these; the contract says 404.
	for _, path := range []string{"/api/nope", "/", "/api/health/", "/api//health", "/api/%68ealth", "/api/x/../health"} {
		response := send(t, "GET", path, sendOptions{})
		assertErrorEnvelope(t, response, 404, "NOT_FOUND")
	}
}

func TestKnownRouteWithTheWrongMethodIsABare405(t *testing.T) {
	wrong := send(t, "DELETE", "/api/health", sendOptions{})
	if wrong.status != 405 || wrong.header.Get("Allow") != "GET,HEAD" || wrong.raw != "" {
		t.Errorf("DELETE /api/health = %d, Allow %q, body %q", wrong.status, wrong.header.Get("Allow"), wrong.raw)
	}

	both := send(t, "PUT", "/api/transcriptions", sendOptions{})
	if both.header.Get("Allow") != "GET,HEAD,POST" {
		t.Errorf("Allow = %q", both.header.Get("Allow"))
	}

	// HEAD is GET without the body, so it is not allowed where GET is not.
	head := send(t, "HEAD", "/api/session", sendOptions{})
	if head.status != 405 || head.header.Get("Allow") != "POST" {
		t.Errorf("HEAD /api/session = %d, Allow %q", head.status, head.header.Get("Allow"))
	}
	if ok := send(t, "HEAD", "/api/health", sendOptions{}); ok.status != 200 {
		t.Errorf("HEAD /api/health = %d", ok.status)
	}
}

func TestEveryAnswerCarriesCORSAndOptionsIsAPreflight(t *testing.T) {
	simple := send(t, "GET", "/api/nope", sendOptions{})
	if simple.header.Get("Access-Control-Allow-Origin") != "*" || simple.header.Get("Access-Control-Expose-Headers") != "*" {
		t.Errorf("CORS headers missing: %v", simple.header)
	}

	preflight := send(t, "OPTIONS", "/api/checkout", sendOptions{headers: map[string]string{
		"Origin":                        "http://localhost:5173",
		"Access-Control-Request-Method": "POST",
	}})
	if preflight.status != 200 || preflight.raw != "" {
		t.Fatalf("preflight = %d %q", preflight.status, preflight.raw)
	}
	for _, name := range []string{"Access-Control-Allow-Origin", "Access-Control-Allow-Methods", "Access-Control-Allow-Headers"} {
		if preflight.header.Get(name) != "*" {
			t.Errorf("%s = %q", name, preflight.header.Get(name))
		}
	}
	// A preflight is answered whether or not the path exists.
	if unknown := send(t, "OPTIONS", "/api/nope", sendOptions{}); unknown.status != 200 {
		t.Errorf("OPTIONS /api/nope = %d", unknown.status)
	}
}

func TestMalformedBodyIs400(t *testing.T) {
	bodies := []string{
		// `duration_seconds` is required, and unknown fields are rejected.
		`{"title":"x","surprise":true}`,
		`{"title":"x"}`,
		`{not json`,
		`[]`,
		`null`,
		``,
		`   `,
		// One JSON value, and nothing after it.
		`{"title":"x","duration_seconds":60} trailing`,
		`{"title":"x","duration_seconds":60}{}`,
		// Keys are matched exactly; encoding/json on its own would accept this one.
		`{"Title":"x","duration_seconds":60}`,
		// Wrong scalar types are type errors, never coerced.
		`{"title":123,"duration_seconds":60}`,
		`{"title":"x","duration_seconds":"60"}`,
		`{"title":"x","duration_seconds":60.5}`,
		`{"title":"x","duration_seconds":true}`,
		`{"title":null,"duration_seconds":60}`,
		`{"title":"x","duration_seconds":null}`,
		`{"title":"x","duration_seconds":2147483648}`,
		`{"title":"x","duration_seconds":1e400}`,
		// Well-typed, but outside the contract's bounds.
		`{"title":"x","duration_seconds":0}`,
		`{"title":"x","duration_seconds":7201}`,
		`{"title":"   ","duration_seconds":60}`,
		`{"title":"` + strings.Repeat("x", 201) + `","duration_seconds":60}`,
	}
	for _, raw := range bodies {
		fake := newFakeMeteroid(nil)
		response := send(t, "POST", "/api/transcriptions", sendOptions{headers: bearer("scribe-demo-1"), body: raw, meteroid: fake})

		if response.status != 400 {
			t.Errorf("%s: status = %d", raw, response.status)
			continue
		}
		assertErrorEnvelope(t, response, 400, "BAD_REQUEST")
		if len(fake.calls) != 0 {
			t.Errorf("Meteroid was called for %s", raw)
		}
	}
}

func TestIntegerSpelledAsAFloatIsStillAnInteger(t *testing.T) {
	// JSON Schema's `type: integer` is about the value, not the spelling. These get past
	// validation — and then fail on the entitlement read, which has no stub here.
	for _, raw := range []string{`{"title":"x","duration_seconds":60.0}`, `{"title":"x","duration_seconds":6e1}`} {
		fake := newFakeMeteroid(nil)
		response := send(t, "POST", "/api/transcriptions", sendOptions{headers: bearer("scribe-demo-1"), body: raw, meteroid: fake})

		if response.status == 400 || len(fake.calls) == 0 {
			t.Errorf("%s was refused: %d %s", raw, response.status, response.raw)
		}
	}
}

func TestBodyThatIsNotValidUTF8Is400(t *testing.T) {
	response := send(t, "POST", "/api/session", sendOptions{body: "{\"\xff\":1}"})
	assertErrorEnvelope(t, response, 400, "BAD_REQUEST")
}

func TestOptionalBodyRejectsWhatItDoesNotDeclare(t *testing.T) {
	for _, raw := range []string{`{"nope":1}`, `{"workspace_name":123}`, `{"workspace_name":"  "}`, `{"email":false}`, `[]`, `null`, `{} {}`} {
		fake := newFakeMeteroid(nil)
		response := send(t, "POST", "/api/session", sendOptions{body: raw, meteroid: fake})

		if response.status != 400 {
			t.Errorf("%s: status = %d", raw, response.status)
			continue
		}
		assertErrorEnvelope(t, response, 400, "BAD_REQUEST")
		if len(fake.calls) != 0 {
			t.Errorf("Meteroid was called for %s", raw)
		}
	}
}

func TestPlanCodeOutsideTheEnumIsRejectedBeforeTheCatalog(t *testing.T) {
	fake := newFakeMeteroid(nil)
	response := send(t, "POST", "/api/checkout", sendOptions{headers: bearer("scribe-demo-1"), body: `{"plan_code":"gold"}`, meteroid: fake})

	assertErrorEnvelope(t, response, 400, "BAD_REQUEST")
	if len(fake.calls) != 0 {
		t.Errorf("Meteroid was called: %v", fake.calls)
	}
}

func TestOutOfRangePortalLifetimeIsRejectedBeforeMeteroid(t *testing.T) {
	for _, raw := range []string{`{"expires_in_seconds":5}`, `{"expires_in_seconds":2592001}`, `{"expires_in_seconds":"60"}`} {
		fake := newFakeMeteroid(nil)
		response := send(t, "POST", "/api/portal-session", sendOptions{headers: bearer("scribe-demo-1"), body: raw, meteroid: fake})

		assertErrorEnvelope(t, response, 400, "BAD_REQUEST")
		if len(fake.calls) != 0 {
			t.Errorf("Meteroid was called for %s", raw)
		}
	}
}

func TestAbsentNullAndEmptyPortalBodiesMeanTheDefault(t *testing.T) {
	for _, raw := range []string{``, ` `, `{}`, `{"expires_in_seconds":null}`} {
		fake := newFakeMeteroid(map[string]stub{
			"POST /api/v1/customers/scribe-demo-1/portal-token": static(map[string]string{"token": "jwt", "portal_url": "https://portal.test"}),
		})
		response := send(t, "POST", "/api/portal-session", sendOptions{headers: bearer("scribe-demo-1"), body: raw, meteroid: fake})

		if response.status != 201 {
			t.Fatalf("%q: %d %s", raw, response.status, response.raw)
		}
		assertJSON(t, response.body, `{"portal_url":"https://portal.test","token":"jwt","expires_in_seconds":86400}`)
		assertJSON(t, string(fake.bodies[0]), `"{\"expires_in_seconds\":86400}"`)
	}
}

func TestInvalidInvoiceLimitIsRejectedBeforeMeteroid(t *testing.T) {
	for _, query := range []string{"limit=0", "limit=101", "limit=abc", "limit=", "limit=1.5", "foo=1", "limit=1&limit=2", "limit=99999999999", "limit=%zz"} {
		fake := newFakeMeteroid(nil)
		response := send(t, "GET", "/api/invoices?"+query, sendOptions{headers: bearer("scribe-demo-1"), meteroid: fake})

		if response.status != 400 {
			t.Errorf("%s: status = %d", query, response.status)
			continue
		}
		assertErrorEnvelope(t, response, 400, "BAD_REQUEST")
		if len(fake.calls) != 0 {
			t.Errorf("Meteroid was called for %s", query)
		}
	}
}

func TestOversizedBodyIsRefusedInTheEnvelope(t *testing.T) {
	response := send(t, "POST", "/api/session", sendOptions{body: strings.Repeat(" ", maxBodyBytes+1)})

	if response.status != 413 {
		t.Fatalf("status = %d", response.status)
	}
	assertJSON(t, response.body, `{"code":"BAD_REQUEST","message":"The request body exceeds the 2097152-byte limit.","quota":null,"upgrade_plan_code":null}`)

	atTheLimit := send(t, "POST", "/api/session", sendOptions{body: strings.Repeat(" ", maxBodyBytes)})
	if atTheLimit.status == 413 {
		t.Error("a body of exactly the limit was refused")
	}
}

func TestUnreachableMeteroidIs502NotACrash(t *testing.T) {
	// No fake here: the real client, pointed at a port nothing listens on.
	unreachable := testConfig
	unreachable.meteroidBaseURL = "http://127.0.0.1:9"
	response := send(t, "GET", "/api/plans", sendOptions{app: newApp(unreachable, unreachable.meteroidClient())})

	assertErrorEnvelope(t, response, 502, "UPSTREAM_ERROR")
}

// ---------------------------------------------------------------- webhooks

func webhook(t *testing.T, payload string, headers map[string]string) testResponse {
	t.Helper()
	return send(t, "POST", "/api/webhooks/meteroid", sendOptions{headers: headers, body: payload})
}

func TestCorrectlySignedEventIsAccepted(t *testing.T) {
	payload := `{"id":"evt_123","type":"invoice.paid","timestamp":"2026-09-01T12:00:00Z"}`
	response := webhook(t, payload, signed(t, payload, "msg_1", "webhook"))

	if response.status != 202 {
		t.Fatalf("%d %s", response.status, response.raw)
	}
	assertJSON(t, response.body, `{"received":true,"event_id":"evt_123","type":"invoice.paid","handled":true}`)
}

func TestRawBytesAreVerifiedNotAReserialization(t *testing.T) {
	// Spacing and key order no JSON serializer would reproduce.
	payload := "{ \"type\" : \"invoice.paid\",\n\t\"id\":\"evt_raw\"   }"
	response := webhook(t, payload, signed(t, payload, "msg_raw", "webhook"))

	if response.status != 202 || response.body["event_id"] != "evt_raw" {
		t.Fatalf("%d %s", response.status, response.raw)
	}
}

func TestUnknownEventTypeIsAcknowledgedNotRejected(t *testing.T) {
	// A receiver that 400s on an unrecognized type breaks the first time Meteroid ships a
	// new one — so this must be a 202 with `handled: false`.
	payload := `{"id":"evt_9","type":"something.invented.later"}`
	response := webhook(t, payload, signed(t, payload, "msg_9", "webhook"))

	if response.status != 202 {
		t.Fatalf("%d %s", response.status, response.raw)
	}
	assertJSON(t, response.body, `{"received":true,"event_id":"evt_9","type":"something.invented.later","handled":false}`)
}

func TestSignedBodyOfAnUnexpectedShapeIsAcknowledgedToo(t *testing.T) {
	for _, payload := range []string{`[1,2]`, `{"id":7,"type":["x"]}`, `"just a string"`, `null`} {
		response := webhook(t, payload, signed(t, payload, "msg_shape", "webhook"))

		if response.status != 202 {
			t.Fatalf("%s: %d %s", payload, response.status, response.raw)
		}
		assertJSON(t, response.body, `{"received":true,"event_id":"msg_shape","type":null,"handled":false}`)
	}
}

func TestEventWithNoIDFallsBackToTheWebhookIDHeader(t *testing.T) {
	payload := `{"type":"subscription.created"}`
	response := webhook(t, payload, signed(t, payload, "msg_42", "webhook"))

	assertJSON(t, response.body, `{"received":true,"event_id":"msg_42","type":"subscription.created","handled":true}`)
}

func TestSvixHeadersAreAcceptedAsAliases(t *testing.T) {
	payload := `{"type":"invoice.finalized"}`
	response := webhook(t, payload, signed(t, payload, "msg_s", "svix"))

	// The id fallback has to follow the alias as well.
	if response.status != 202 || response.body["event_id"] != "msg_s" {
		t.Fatalf("%d %s", response.status, response.raw)
	}
}

func TestBodyThatDoesNotMatchItsSignatureIsRejected(t *testing.T) {
	headers := signed(t, `{"id":"evt_1","type":"invoice.paid"}`, "msg_1", "webhook")
	// One field different from what was signed.
	response := webhook(t, `{"id":"evt_1","type":"invoice.void"}`, headers)

	assertErrorEnvelope(t, response, 400, "WEBHOOK_SIGNATURE_INVALID")
}

func TestStaleAndFutureTimestampsAreRejected(t *testing.T) {
	payload := `{"type":"invoice.paid"}`
	for _, skew := range []time.Duration{-time.Hour, time.Hour} {
		headers := signed(t, payload, "msg_old", "webhook")
		headers["webhook-timestamp"] = strconv.FormatInt(time.Now().Add(skew).Unix(), 10)

		assertErrorEnvelope(t, webhook(t, payload, headers), 400, "WEBHOOK_SIGNATURE_INVALID")
	}
}

func TestUnsignedEventIsRejected(t *testing.T) {
	response := webhook(t, `{"type":"invoice.paid"}`, map[string]string{"Content-Type": "application/json"})
	assertErrorEnvelope(t, response, 400, "WEBHOOK_SIGNATURE_INVALID")
}

func TestCorrectlySignedBodyThatIsNotJSONIsRejected(t *testing.T) {
	for _, payload := range []string{"not json", "", "{} {}", "\xff"} {
		response := webhook(t, payload, signed(t, payload, "msg_nj", "webhook"))
		if response.status != 400 {
			t.Errorf("%q: status = %d", payload, response.status)
			continue
		}
		assertErrorEnvelope(t, response, 400, "WEBHOOK_SIGNATURE_INVALID")
	}
}

func TestMissingOrUnusableWebhookSecret(t *testing.T) {
	// Missing is the operator's problem, not a bad signature.
	missing := testConfig
	missing.meteroidWebhookSecret = ""
	response := send(t, "POST", "/api/webhooks/meteroid", sendOptions{app: newApp(missing, newFakeMeteroid(nil).client), body: "{}"})
	assertErrorEnvelope(t, response, 500, "INTERNAL")

	// One that is not base64 can verify nothing, which is how the other backends see it.
	garbled := testConfig
	garbled.meteroidWebhookSecret = "whsec_!!!"
	response = send(t, "POST", "/api/webhooks/meteroid", sendOptions{app: newApp(garbled, newFakeMeteroid(nil).client), body: "{}"})
	assertErrorEnvelope(t, response, 400, "WEBHOOK_SIGNATURE_INVALID")
}

func TestAPanicIsTheInternalEnvelopeNotADroppedConnection(t *testing.T) {
	// A nil Meteroid client is a bug by construction: the first SDK call dereferences it.
	response := send(t, "GET", "/api/plans", sendOptions{app: newApp(testConfig, nil)})

	assertErrorEnvelope(t, response, 500, "INTERNAL")
	if message := fmt.Sprint(response.body["message"]); message != "Unexpected error handling the request." {
		t.Errorf("the panic leaked into the message: %s", message)
	}
}
